### ISSUE TITLE
Option to delete/change associated data after eBook file is deleted or moved

### ISSUE THREAD

**Issue by itprojects** [johnfactotum/foliate#73]
Books can be opened, but not closed, and thus not removed from the app.

Details:

1. Open Foliate, pick an epub A.epub loads correctly if it exists

2. Close Foliate

3. Rename A.epub file to B.epub

4. Open Foliate: home screen has two options - Open, Continue Reading

5. Press Continue Reading: "Oh no!, The file cannot be opened"

Expected behaviour: if the file cannot be found then ask the user to remove it from the internal database (dconf gschema value com.github.johnfactotum.Foliate.last-file), or to replace it with the correct file name and location (keeping one's bookmarks etc).

**Comment by johnfactotum** [comment]
I don't think it would be possible to detect the new filename without some kind of file indexing like Tracker, which is something I'm not planning to add.

Note that bookmarks are stored based on the book's identifier (defined in the metadata), not filename, so if you just open B.epub (from the file manager, or the file chooser dialog), everything should be there.

I suppose I'll add a check to see if the file exists before offering to continue reading it.

**Comment by itprojects** [comment]
When a person finishes an ebook they will probably delete the epub.
This bug is all about the user's choice of action to be taken.
For example:

1. delete notes and bookmarks
2. keep notes and bookmarks, file is relocated (File-URL-Picker Dialog)
3. delete all database data about this book

It's great that you've decided to avoid the Tracker. Gnome-books fell into the trap of using the Tracker as a default and only way of viewing books. They're still in it! Without a tracker no books appear in the library, and no books can be opened separately! There is even an (ignored) bug filed for this.

**Comment by johnfactotum** [comment]
Okay, I think I kind know what you mean... It's really a separate issue from restoring the last file.

I suppose it may be possible to associate URIs with each identifier. And then you can, for example, do a sort of cleanup by purging all data associated with non-existent URIs, or choose a new URI where the new fileis now. Is this the kind use case you have in mind?

I don't really like this. The main point of the current system is precisely to a to avoid any databases, any scanning of folders, and any reference to files as much as possible. Otherwise, it would be much easier to just use something like Tracker or GVfs metadata. I think implementing this would be verging towards a full-blown library management feature (that is, to have this, you'd basically have to build the basic library infrastructure, whether or not more advanced management features will be there or not)

That being said, I do think it would be good (or even necessary) to purge cache, notes and bookmarks for books you no longer have. An easy workaround for the user would be to simply delete the data before deleting the files. (Renaming is not relevant, because, as I said, the data aren't associated with URIs, but rather identifiers)

**Comment by itprojects** [comment]
Yes, you're absolutely right. Best to avoid the library entirely. Just giving an option to remove (?export to txt file in some formatted log-output kind of way) the data should be just fine. Nothing heavy.

**Comment by itprojects** [comment]
Possibly a New Bug:

1. Open Foliate:
        Pick an epub A.epub loads correctly if it exists

2. Close Foliate

3. Delete A.epub

4. Open Foliate again:

        Home screen has two options - Open; Continue Reading

5. Press Continue Reading:
         "Oh no!, The file cannot be opened"

6. You have to press 'Open...' from the headerbar menuitem.

        This opens a file picker Dialog, where you find and select B.epub

7. Click 'Open':
        Dialog closes **A Second Foliate appears**

8. If you close the second Foliate first, and only then the first Foliate, you get a bug!

The new file IS NOT the newly user picked B.epub file in the second Foliate.

However if you close the first Foliate and only then open the second Foliate, the bug doesn't happen.

Note: Dconf seems to overwrite from RAM the last-current wrong epub URL.

**Comment by johnfactotum** [comment]
That is not a bug. That is the expected behavior. It's restoring the last closed file, not the last opened file.

**Comment by itprojects** [comment]
Why should there be two Foliate windows? The dialog is understandable, just not the opened second Foliate window. Surely, the second one should have closed the first 'Ooops'-ied window, then start the new one. Perhaps that's a quick fix?

**Comment by johnfactotum** [comment]
The current behavior is not that unreasonable. Evince also behaves the same way. And it should most definitely not close the window before starting the new one. If anything, it should maybe open in the very same window.

**Comment by itprojects** [comment]
Yes, opening in the same window would do, thus overriding last read book.

**Comment by itprojects** [comment]
This bug is now partly solved.

When the book is deleted, the option to open another is correctly given.
This leaves the multiple-window riddle to be solve.

1. Should there be only one instance of Foliate?

2. Should the 'Open' menuitem be allowed to open another Foliate window?
If yes, then whichever Foliate window is last closed, assigns the book to be opened next time.
If no, then menuitem 'Open' is causing a logical bug, by opening the second window.

**Comment by johnfactotum** [comment]
Foliate _is_ a multi-window application, and what you're describing ("whichever Foliate window is last closed, assigns the book to be opened next time") is exactly how it works now.

**Comment by itprojects** [comment]
So then we can close this bug, since data export is part of https://github.com/johnfactotum/foliate/issues/93.

**Comment by anishmittal2020** [comment]
@itprojects Is it possible to clear the library history?

**Comment by itprojects** [comment]
@anishmittal2020 You can clear the library history only manually.

You would look in the Foliate folder and delete the cache, here's how.

1. If you have a native package repository like `.deb` or `.rpm`

Go to folder:

/home/USERNAME/.cache/com.github.johnfactotum.Foliate

/home/USERNAME/.local/share/com.github.johnfactotum.Foliate

Delete the relevant files in that folder.

2. If you have the `.flatpak`

Go to folder:

/home/USERNAME/.var/app/com.github.johnfactotum.Foliate/cache

/home/USERNAME/.var/app/com.github.johnfactotum.Foliate/data

Delete the relevant files in that folder.

**BE WARNED** - this will delete any annotations or bookmarks you have in Foliate.

Maybe your question is the topic for a Feature Request, bug _probably_ not a closed bug thread...

**Comment by johnfactotum** [comment]
One can delete data for individual books in the library window.

For cleaning data for deleted books, see #582.
